KAFKA-16717 [3/N]: Add AdminClient.alterShareGroupOffsets

JimmyWang6 opened the pull request:
[KAFKA-16717](https://issues.apache.org/jira/browse/KAFKA-16717) aims to
finish the AlterShareGroupOffsets for ShareGroupCommand part.

Reviewers: Lan Ding <isDing_L@163.com>, Chia-Ping Tsai
 <chia7712@gmail.com>, TaiJuWu <tjwu1217@gmail.com>, Andrew Schofield
 <aschofield@confluent.io>

JimmyWang6 commented:
Hi all @AndrewJSchofield @chia7712 @smjn @apoorvmittal10 ，
I have finished the basic part of `kafka-share-groups.sh` --reset-offsets`. PTAL if you have time ;)

JimmyWang6 commented:
Hi @AndrewJSchofield @smjn @apoorvmittal10 ,
Sorry for the late response, could you please take a look if there is anything wrong with the current approach? Thanks in advance.

AndrewJSchofield commented:
@JimmyWang6 There's a build failure due to spotless check. Please resolve it and merge latest changes.

AndrewJSchofield commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +428,70 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+            printOffsets(groupOffsets, opts.options.has(opts.verboseOpt));
+        }
+
+        private Map<TopicPartition, OffsetAndMetadata> prepareOffsetsToReset(String groupId) throws ExecutionException, InterruptedException {
```
The error handling for `bin/kafka-share-groups.sh --bootstrap-server localhost:9092 --group G1 --reset-offsets --dry-run` needs to be stronger. I didn't specify any topic information so the map `offsetsByTopicPartitions` is empty and the command blows up.

JimmyWang6 replied:
Good catch. I’ve fixed the bug by adding validation to ensure that either --all-topics or --topic is provided.

AndrewJSchofield commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +428,70 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+        }
+
+        private Map<TopicPartition, OffsetAndMetadata> prepareOffsetsToReset(String groupId) throws ExecutionException, InterruptedException {
+            Map<String, ListShareGroupOffsetsSpec> groupSpecs = new HashMap<>();
```
This probably ought to be `groupSpecs = Map.of(groupId, new ListShareGroupOffsetsSpec())`.

JimmyWang6 replied:
Good catch, thanks.

AndrewJSchofield commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +428,70 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+            Map<TopicPartition, OffsetAndMetadata> offsetsByTopicPartitions = adminClient.listShareGroupOffsets(groupSpecs).all().get().get(groupId);
+
+            if (opts.options.has(opts.topicOpt)) {
+                Set<String> topics = new HashSet<>(opts.options.valuesOf(opts.topicOpt));
```
Maybe `Set.copyOf(opts.options.valuesOf(opts.topicOpt))`

JimmyWang6 replied:
Good idea.Updated.

AndrewJSchofield reviewed (changes requested):
Thanks for the updated PR.

The overall approach of obtaining the offsets seems about right. It needs a lot of work to get the behaviour neat and tidy in all of the various cases it can be used in (such as non-existent topic, no topics specified, list of topics specified, incorrect arguments, and so on).

Even with a couple of local changes, I was not able to get the offsets to reset, but I think it's close.

Please continue in this direction and I'll continue to review as the changes appear. Thanks.

chia7712 commented:
@JimmyWang6 could you please fix the build error?

JimmyWang6 commented:
Hi @chia7712，
Thanks for your comment and I have updated the PR, PTAL if you have time.

JimmyWang6 commented:
Hi @AndrewJSchofield  @chia7712 
I have updated the PR and resolved several bugs.
Here are some of my test scenarios：

1. When a single shareConsumer subscribes to a single topic, consumes some messages, and then exits and , it can subsequently consume all previously consumed messages again after resetting the offset to ealiest.
2. When a single shareConsumer subscribes to mutiple topics, then excute `bin/kafka-share-groups.sh --reset-offsets --execute --to-earliest --all-topics --group test-group --bootstrap-server localhost:9092 `, the offset of all the topics were reset.
3. When a single shareConsumer subscribes to mutiple topics, consumes some messages, and then exits, it can subsequently consume all previously consumed messages again after resetting the consumption offset.
4. If no topic argument is provided when running `bin/kafka-share-groups.sh --bootstrap-server localhost:9092 --group test-group --reset-offsets --dry-run`, the following exception occurs: Option [reset-offsets] requires one of these options: [all-topics] or [topic].
5. When none of `to-datetime`, `to-earliest` or `to-latest` are specified, the execution failed with an error `Option '[reset-offsets]' requires one of the following scenarios: [[to-datetime], [to-earliest], [to-latest]]`
6. For none-exists topic, the execution failed with `Some topics none do not exist in share group 'test-group'.`

chia7712 commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -3772,7 +3772,19 @@ class KafkaApis(val requestChannel: RequestChannel,
           case Some(error) =>
             topic.partitions().forEach(partition => responseBuilder.addPartition(topic.topicName(), partition.partitionIndex(), metadataCache.topicNamesToIds(), error.error))
           case None =>
-            authorizedTopicPartitions.add(topic)
```
Excuse me, why to make this change? it makes a copy of `topic`

JimmyWang6 replied:
Hi @chia7712 ,
Much thanks for your review. 
This is a bug related to my previous PR :( The add method always returns false and fails to add the topic correctly because the condition `newElement.prev() != INVALID_INDEX || newElement.next() != INVALID_INDEX `evaluates to true. I believe this occurs because the topic has already been added to an `ImplicitLinkedHashCollection`, meaning the prev() and next() pointers of the specified topic are no longer invalid. Therefore, I'm attempting to create a new topic here.

    public final boolean add(E newElement) {
        if (newElement == null) {
            return false;
        }
        if (newElement.prev() != INVALID_INDEX || newElement.next() != INVALID_INDEX) {
            return false;
        }
        if ((size + 1) >= elements.length / 2) {
            changeCapacity(calculateCapacity(elements.length));
        }
        int slot = addInternal(newElement, elements);
        if (slot >= 0) {
            addToListTail(head, elements, slot);
            size++;
            return true;
        }
        return false;
    }

chia7712 replied:
Thanks for explaining. How about using `authorizedTopicPartitions.add(topic.duplicate())` instead?

JimmyWang6 replied:
Good idea! Updated.

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -682,15 +682,43 @@ CompletableFuture<AlterShareGroupOffsetsResponseData> persisterInitialize(
+                        .setTopicName(topic.topicName());
+                    topic.partitions().forEach(partition -> {
+                        AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponsePartition partitionData;
+                        PartitionErrorData error = result.getErrors().get(topic.topicId()).get(partition.partitionIndex());
```
`result.getErrors()` will create a new collection by grouping data. Perhaps it should be moved out of loop.

chia7712 replied:
Furthermore, how do we ensure that the topic from `AlterShareGroupOffsetsResponseData` must exist in `InitializeShareGroupStateResult`?

If users attempt  to alter the nonexistent partitions within an existing topic, the `AlterShareGroupOffsetsResponseData` will contain the topic but `InitializeShareGroupStateResult` will not. Is this correct?

JimmyWang6 replied:
Good catch! My idea is if users want to alter one nonexistent partition, the error code and message will be added to `AlterShareGroupOffsetsResponseData` when executing `groupMetadataManager.completeAlterShareGroupOffsets`. So I plan to skip the processed topicPartition which already had an error code, for which the rest of the partitions should exist in InitializeShareGroupStateResult.

chia7712 reviewed (commented):
@JimmyWang6 thanks for this patch

DL1231 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +378,71 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+                    .filter(entry -> {
+                        TopicPartition topicPartition = entry.getKey();
+                        return topics.contains(topicPartition.topic());
+                    })
```
```suggestion
                    .filter(entry -> topics.contains(entry.getKey().topic()))
```

JimmyWang6 replied:
Updated. Thanks for your review.

DL1231 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -682,15 +682,48 @@ CompletableFuture<AlterShareGroupOffsetsResponseData> persisterInitialize(
+                })
+                .iterator()));
+        return data;
+        
```
nit: extra blank

JimmyWang6 replied:
Fixed.

DL1231 reviewed (approved):
Overall LGTM, left some small comments.

AndrewJSchofield commented:
@JimmyWang6 I've spent a while trying out this code and I wasn't able to get the majority of the error scenarios to work as I expected. The problem I think is the `Admin.alterShareGroupOffsets` method which I don't think was handling top-level errors. I've put in a separate PR https://github.com/apache/kafka/pull/20049 for that which I hope will help things along.

AndrewJSchofield commented:
@JimmyWang6 Please merge latest changes. Now that https://github.com/apache/kafka/pull/20049 is merged, the `Admin.alterShareGroupOffsets` will now return top-level errors correctly. This means `kafka-share-groups.sh --reset-offsets` in error cases will fail, rather than appearing to work when it did not.

JimmyWang6 commented:
@AndrewJSchofield So much thanks for your help to resolving the issues. I sincerely apologize for the previous bugs I introduced. I’ll resolve the conflicts and thoroughly test the script as soon as possible to ensure everything works smoothly.

JimmyWang6 commented:
Hi @chia7712 @AndrewJSchofield ,
I have updated the PR and addressed the [comments](https://github.com/apache/kafka/pull/18929#pullrequestreview-3002388120), PTAL when get a chance, thanks.

AndrewJSchofield commented:
Thanks for the updates. I'm starting to review and test it. Here's a small problem I've found so far. `SG1` is an existing share group with 1 member. `T2` doesn't exist, so the message about "Not subscribed" is fine. However, the usage message for the tool is printed even though it's a regular expected error.

```
bin/kafka-share-groups.sh --bootstrap-server localhost:9092 --reset-offsets --group SG1 --execute --topic T2 --to-earliest 
Share group 'SG1' is not subscribed to topic 'T2'
Option                                 Description                           
------                                 -----------                           
--all-groups                           Apply to all share groups.
(( And so on ))
```

AndrewJSchofield commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +378,67 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+            Map<String, ShareGroupDescription> shareGroups = describeShareGroups(Collections.singleton(groupId));
+            TreeMap<String, Entry<ShareGroupDescription, Collection<SharePartitionOffsetInformation>>> groupOffsets = new TreeMap<>();
+            shareGroups.forEach((id, description) -> groupOffsets.put(id, new SimpleImmutableEntry<>(description, partitionOffsets)));
+            printOffsets(groupOffsets, opts.options.has(opts.verboseOpt));
```
The `--verbose` option is not valid with `--reset-offsets` so this is a bit weird. But actually, I think it would be better if there was something like the consumer group tool `printOffsetsToReset`. This is just because I think having a column heading `NEW-START-OFFSET` would make it clear that it's the new start offset.

JimmyWang6 replied:
Good idea! I will use the same format of `printOffsetsToReset` of consumer group tool.

AndrewJSchofield commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +378,67 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+                    .collect(Collectors.toSet());
+                if (!existsTopics.containsAll(topics)) {
+                    CommandLineUtils
+                        .printUsageAndExit(opts.parser, String.format("Share group '%s' is not subscribed to topic '%s'",
```
This doesn't distinguish between non-existent topics and topics which are just not subscribed. This is probably a debatable point, so I wonder what you think.

JimmyWang6 replied:
I agree with your point. I've already added validation for non-existent topics, which require an extra `adminClient.listTopics()` request.

AndrewJSchofield reviewed (changes requested):
Thanks for the updates. This is much closer now. It looks to me like it's working properly when it resets the offsets.

There's quite a lot of tidying up to do for the error cases. Here are some I tried and it was a bit messy:

* Attempt to reset offsets for a consumer group
* Attempt to reset offsets for a non-existent group
* Attempt to reset offsets when there are members in the share group
* Attempt to reset offsets for non-existent topic
* Attempt to reset offsets for existing topic which is not subscribed in the share group

I think the code structure will change a little when you sort out the error handling. Too often, we can see a stack trace when, for expected errors, a simple message would be neater. Also, the usage message sometimes comes out when the command options were valid.

AndrewJSchofield commented:
Please fix the checkstyle build error.

JimmyWang6 commented:
Hi @AndrewJSchofield ,
I have improved the error handling structure, please take another look when get a chance, thank! Here are my test results:

- Attempt to reset offsets for a consumer group

<img width="1425" height="137" alt="image" src="https://github.com/user-attachments/assets/484164d3-2d2e-4b90-af09-c925e59d41f5" />

- Attempt to reset offsets for a non-existent group

<img width="1419" height="149" alt="image" src="https://github.com/user-attachments/assets/3293ece3-24fc-4ee3-be5d-4a2b64a3b63f" />

- Attempt to reset offsets when there are members in the share group

<img width="1158" height="122" alt="image" src="https://github.com/user-attachments/assets/5e8bd4ab-b13b-430d-9187-96368bef1880" />


- Attempt to reset offsets for non-existent topic

<img width="1415" height="109" alt="image" src="https://github.com/user-attachments/assets/499a4336-c08e-4267-9bf1-603da3fd45fd" />

- Attempt to reset offsets for existing topic which is not subscribed in the share group

<img width="1376" height="85" alt="image" src="https://github.com/user-attachments/assets/90c9267b-9f6b-4876-806f-4bc016b4c057" />

- Successfully to reset offsets to earliest

<img width="1428" height="116" alt="image" src="https://github.com/user-attachments/assets/873525f4-1189-462e-b538-e235df0aec68" />

AndrewJSchofield commented:
Still getting the usage message for trying to reset offsets for a consumer group.

```
bin/kafka-share-groups.sh --bootstrap-server localhost:9092 --reset-offsets --group CG1 --topic T1 --execute --to-earliest
Group CG1 is not a share group.
Option                                 Description                           
------                                 -----------                           
--all-groups                           Apply to all share groups.            
--all-topics                           Consider all topics assigned to a     
                                         share group in the 'reset-offsets'  
                                         process
```

AndrewJSchofield commented on tools/src/main/java/org/apache/kafka/tools/OffsetsUtils.java:
```diff
@@ -69,6 +69,21 @@ public OffsetsUtils(Admin adminClient, OptionParser parser, OffsetsUtilsOptions
     }
 
+    public static void printOffsetsToReset(Map<String, Map<TopicPartition, OffsetAndMetadata>> groupAssignmentsToReset) {
+        String format = "%n%-30s %-30s %-10s %-15s";
```
If you look at `ShareGroupCommand.printOffsetsFormat`, you'll see that the column alignment is much neater. I suggest:
* group is  `max(15, groupId.length())` characters wide
* topic is `max(15, maxTopicLen)` characters wide
* partition is 10 characters wide
* offset does not specify

JimmyWang6 replied:
Copy that.

AndrewJSchofield commented on tools/src/test/java/org/apache/kafka/tools/consumer/group/ShareGroupCommandTest.java:
```diff
@@ -1052,6 +1059,217 @@ public void testDeleteShareGroupsPartialFailure() {
     }
+    
+    @Test
+    public void testAlterShareGroupAllTopicsSuccess() {
```
This is testing multiple topics, not `--all-topics`. Both are valid.

AndrewJSchofield commented on tools/src/test/java/org/apache/kafka/tools/consumer/group/ShareGroupCommandTest.java:
```diff
@@ -1052,6 +1059,217 @@ public void testDeleteShareGroupsPartialFailure() {
+        String group = "share-group";
+        String topic = "topic";
+        String bootstrapServer = "localhost:9092";
+        String[] cgcArgs = new String[]{"--bootstrap-server", bootstrapServer, "--reset-offsets", "--to-earliest", "--dry-run", "--topic", topic, "--group", group};
```
Let's have some tests for `--to-latest` and `--to-datetime`.

JimmyWang6 replied:
I have added `testAlterShareGroupToLatestSuccess()` and `testAlterShareGroupAllTopicsToDatetimeSuccess()` tests for those two cases.

AndrewJSchofield reviewed (changes requested):
Thanks for the update.

In general, I think it's working. The error cases are also working as I expect, but it is generally the case that the usage string is being printed when an expected error occurs and this is not desirable.

Compare `kafka-consumer-groups.sh`
```
% bin/kafka-consumer-groups.sh --bootstrap-server localhost:9092 --reset-offsets --group SG1 --dry-run --to-latest --topic T1 --topic NONEXISTENT 

Error: Executing consumer group command failed due to This server does not host this topic-partition.
org.apache.kafka.common.errors.UnknownTopicOrPartitionException: This server does not host this topic-partition.
```
and `kafka-share-groups.sh`
```
% bin/kafka-share-groups.sh --bootstrap-server localhost:9092 --reset-offsets --group SG1 --dry-run --to-latest --topic T1 --topic NONEXISTENT
Topic NONEXISTENT not present in metadata.
Option                                 Description                           
------                                 -----------                           
--all-groups                           Apply to all share groups.            
--all-topics                           Consider all topics assigned to a     
                                         share group in the 'reset-offsets'  
                                         process.                       
```

You should only get the usage message if there's a problem with the command line arguments.

AndrewJSchofield commented:
Please also merge latest changes from trunk. I expect the build will be clean afterwards.

JimmyWang6 commented:
> Still getting the usage message for trying to reset offsets for a consumer group.
> 
> ```
> bin/kafka-share-groups.sh --bootstrap-server localhost:9092 --reset-offsets --group CG1 --topic T1 --execute --to-earliest
> Group CG1 is not a share group.
> Option                                 Description                           
> ------                                 -----------                           
> --all-groups                           Apply to all share groups.            
> --all-topics                           Consider all topics assigned to a     
>                                          share group in the 'reset-offsets'  
>                                          process
> ```

Oops, I'm sorry that I didn't grasp your meaning, I will try to replace `CommandLineUtils.printUsageAndExit()` method call by a simple message.

JimmyWang6 committed:
address comments

JimmyWang6 committed:
remove test code

JimmyWang6 committed:
more tests for share group command

JimmyWang6 commented:
@AndrewJSchofield I have addressed the below comments as your suggestions. Please take a glance to see if they look good. Thanks! :)

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +378,75 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+            String groupId = opts.options.valueOf(opts.groupOpt);
+            try {
+                ShareGroupDescription shareGroupDescription = describeShareGroups(List.of(groupId)).get(groupId);
+                if (!GroupState.EMPTY.equals(shareGroupDescription.groupState())) {
```
Should `DEAD` state be included?

JimmyWang6 replied:
I thinks we should, fixed.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +378,75 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+            Map<TopicPartition, OffsetAndMetadata> offsetsByTopicPartitions = adminClient.listShareGroupOffsets(groupSpecs).all().get().get(groupId);
+            Set<String> existsTopics = adminClient.listTopics().names().get();
+
+            if (opts.options.has(opts.topicOpt)) {
```
This behavior is not consistent with normal consumer tool. It should include following tasks.

1. parse the string for "topic" and "topic partition". It means both `--topicc a` and `--topic a:0` are valid
2. parse the `allTopicsOpt`
3. if `allTopicsOpt` is defined, all topics in the group should be reset.
4. if `topicOpt` is defined, all specified tps should be reset

AndrewJSchofield replied:
Yes, this is a good point. You cannot delete the offset for individual topic-partitions. You can reset for individual topic-partitions.

JimmyWang6 replied:
@chia7712 Thanks for your comments, I will handle the topic-partions case.

JimmyWang6 replied:
@chia7712 @AndrewJSchofield 
I have modified the current implementation by reusing offsetsUtils.parseTopicPartitionsToReset(). However, when attempting to pass a non-existent topic-partition (e.g., topicA:999), a TimeoutException is thrown as shown below:
`Caused by: java.util.concurrent.ExecutionException: 
org.apache.kafka.common.errors.TimeoutException: 
Call(callName=listOffsets(api=METADATA), deadlineMs=1753150292432, tries=73308, nextAllowedTryMs=1753150293436) 
timed out at 1753150292436 after 73308 attempt(s)`
This behavior is identical when resetting offsets for a non-existent topic-partition for a consumer group. I think maybe we could create a new issue to address this scenario, as the current implementation does not handle invalid topic-partitions gracefully. I wonder what you think?

chia7712 replied:
@JimmyWang6 yes, that was a issue in `kafka-consumer-groups.sh` and it was already fixed by @xijiu see #20168 #20064

you could refactor the code to ensure the fixes could be applied on `ShareGroupCommand` too.

JimmyWang6 replied:
Noted.

JimmyWang6 replied:
I need to move `checkAllTopicPartitionsValid`, `filterNonExistentPartitions` and `filterNoneLeaderPartitions` method to OffsetUtils to reuse the verification logic for consumer groups. I just noticed there's a remaining issue with that (#20168). I think I could apply the fixes for shared groups after that issue is resolved. Please help to take another look if you get chance.

AndrewJSchofield replied:
@JimmyWang6 I see that the issue in #20168 is still occurring. I'm not sure that you need PR #20235 to merge before you can address the problem of a long delay if resetting offsets for a non-existent partition.

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -709,15 +709,47 @@ CompletableFuture<AlterShareGroupOffsetsResponseData> persisterInitialize(
+                .map(topic -> {
+                    AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponseTopic topicData = new AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponseTopic()
+                        .setTopicName(topic.topicName());
+                    Map<Uuid, Map<Integer, PartitionErrorData>> topicPartitionErrorsMap = result.getErrors();
```
Could you please avoid creating the `topicPartitionErrorsMap` for each loop (`topic`)?

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -709,15 +709,47 @@ CompletableFuture<AlterShareGroupOffsetsResponseData> persisterInitialize(
+            new AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponseTopicCollection(response.responses().stream()
+                .map(topic -> {
+                    AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponseTopic topicData = new AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponseTopic()
+                        .setTopicName(topic.topicName());
```
not sure why the topic id is not set for the response

JimmyWang6 replied:
Thanks for pointing it out, this is a little mistake and I have addressed it.

TaiJuWu replied:
@JimmyWang6 Do you fix this issue or I missed something ?

JimmyWang6 replied:
Sorry, I accidentally missed this issue earlier, but it has now been fixed. Thanks for reminding me!

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -709,15 +709,47 @@ CompletableFuture<AlterShareGroupOffsetsResponseData> persisterInitialize(
+                            return;
+                        }
+                        AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponsePartition partitionData;
+                        PartitionErrorData error = topicPartitionErrorsMap.get(topic.topicId()).get(partition.partitionIndex());
```
should we check the null for `topicPartitionErrorsMap.get(topic.topicId())`?

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -709,15 +709,47 @@ CompletableFuture<AlterShareGroupOffsetsResponseData> persisterInitialize(
+                        AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponsePartition partitionData;
+                        PartitionErrorData error = topicPartitionErrorsMap.get(topic.topicId()).get(partition.partitionIndex());
+                        if (error == null) {
+                            partitionData = new AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponsePartition()
```
we could reuse the `partition` instead of creating new `AlterShareGroupOffsetsResponsePartition`, right?

JimmyWang6 replied:
Good idea, I have change to use partition.duplicate() to build the response.

chia7712 reviewed (commented):
@JimmyWang6 thanks for updates. a couple of comments are left. PTAL

JimmyWang6 committed:
address comments

JimmyWang6 committed:
address comments

TaiJuWu commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ShareGroupCommand.java:
```diff
@@ -366,6 +378,74 @@ Entry<Throwable, Map<String, Throwable>> sendDeleteShareGroupOffsetsRequest(Stri
+
+        protected Map<TopicPartition, OffsetAndMetadata> prepareOffsetsToReset(String groupId) throws ExecutionException, InterruptedException {
+            Map<String, ListShareGroupOffsetsSpec> groupSpecs = Map.of(groupId, new ListShareGroupOffsetsSpec());
+            Map<TopicPartition, OffsetAndMetadata> offsetsByTopicPartitions = adminClient.listShareGroupOffsets(groupSpecs).all().get().get(groupId);
```
What happened if we can't get response from broker?
Should we add try catch or error handler here?

JimmyWang6 replied:
Thanks for your comments. I have added a try-catch block at the outer `resetOffsets()` layer. If it's a KafkaException, the errorMessage will be printed directly. For any other exceptions, they will be thrown directly.

JimmyWang6 committed:
resolve conflicts

AndrewJSchofield reviewed (changes requested):
Thanks for the updates. It's looking much closer now. Local testing works as expected with the exception of a long delay for a non-existent partition. I think we have a fix for that in the consumer group tool that would apply here.

JimmyWang6 committed:
apply fix

JimmyWang6 commented:
@AndrewJSchofield I've applied the fixes from both PRs and the long delay issue is resolved. However, there might be conflicts with the current changes cc @TaiJuWu.

AndrewJSchofield reviewed (approved):
Thanks for the diligent work on this PR. Looks good to me.

JimmyWang6 commented:
Cheers~
